File: Data Structures/Stacks and Queues/AJs/stack.py
# last in, first out -> push, pop would be more time for pop, so shift, unshift
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
    
    def __repr__(self):
        return self.__str__()
        return "Node()"

    def __str__(self):
        return str(self.__class__) + ": " + str(self.__dict__)

class Stack:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    def get(self, index: int):
        if index > (self.size-1): return -1
        count = 0
        current = self.first
        while count<index:
            current = current.next
            count+=1
        return current.value

    def __str__(self):
       return str(self.__class__) + ": " + str(self.__dict__)
    
    def push(self, value): # adds to the beginning
        node = Node(value)
        if self.first is None:
            self.first = node
            self.last = node
        else:
            node.next = self.first
            self.first = node
        self.size+=1
        return self

    def pop(self): # removes from the beginning
        if self.size == 0: return None
        first = self.first
        if self.size == 1:
            self.first = None
            self.last = None
        else:
            self.first = self.first.next
        self.size-=1
        return first.value

    def print(self):
        current = self.first
        while current:
            print(current.value)
            current = current.next
        print('Size is:', self.size)
stack = Stack()
stack.push('1')
stack.push('2')
stack.push('3')
stack.push('4')
stack.push('5')
        

File: Data Structures/Python_JS - Data Structures/Stacks and Queues/QJs/stack.js

class Node {
    constructor(value) {
        this.value = value
        this.next = null
    }    
}

class Stack {
    constructor() {
        this.first = null
        this.last = null
        this.size = 0
    }

    push(value) {
        let node = new Node(value)
        if(this.size == 0) {
            this.first = node
            this.last = node
        } else {
            let currentHead = this.first
            this.first = node
            this.first.next = currentHead
        }
        return ++this.size
    }

    pop() {
        if(!this.first) return null
        let oldhead = this.first

        if(this.size==1) {
            this.first = null
            this.last = null
        } else {
            this.first = this.first.next
        }
        this.size--

        return oldhead.value
    }
}


let stack = new Stack()
stack.push('First')
stack.push("Second")
stack.push("Third")

log = (val) => console.log(val)
log(stack.pop())
log(stack.push('Hello there!'))
log(stack)